test/Mono.Linker.Tests.Cases/DataFlow/MethodReturnParameterDataFlow.cs

```diff
@@ -189,6 +191,10 @@ Type ReturnWithRequirementsAlwaysThrows ()
 			throw new NotImplementedException ();
 		}
 
+		[ExpectedWarning ("IL2106", nameof (UnsupportedReturnType))]
```

Comment:
This and #2048 make me wonder how much do we gain from keeping both `ExpectedWarning` and `URAP` -- I get that `URAP` is more granular and only produced in very specific scenarios, but can't we just add a mechanism for keeping the origin of the produced warnings when testing and then just use ExpectedWarning` everywhere? Furthermore, I think this could allow us get rid of `ReflectionPatternContext` (letting us simplify some of the code in ReflectionMethodBodyScanner and the test infra).
Thinking out loud here, since there might be something I'm completely ignoring.

Reply:
You're not wrong at all. The `IReflectionPatternRecorder` should be removed. I added it a long time ago when the DataFlow work started, but the design has changed since. Its original purpose was to separate the analysis from reporting. We thought that reporting the warnings would be a complex piece of logic (bubbling warnings up the virtual call stacks and so on).

Currently it's basically useless - the only implementation is the logger, which doesn't need that complexity, and the test infra - which can migrate to `ExpectedWarning` just like you describe.